### PR TITLE
Homework, Day 4

### PR DESCRIPTION
**Pull request by harmanssingh**

### CODE REVIEW

File: D04_ex00.py [hunk 1]
```diff
@@ -11,19 +11,40 @@
...
+
+for count in range(5):
+    try:
+        user_inp = int(input("Guess the number: "))
+        if user_inp == rand_num:
+            print("Great! You guessed it right")
+            break
+        elif user_inp > rand_num and user_inp <= 25:
+            print("Too high")
+            continue
```
Comment (danielsgriffin) [inline D04_ex00.py]: Are these continues necessary? Helpful?

File: HW04_ch07_ex02.py [hunk 2]
```diff
@@ -21,15 +21,21 @@
 
 ###############################################################################
 # Imports
-
+import math
 
 # Body
-
+def eval_loop():
+    inputexpression=''
```
Comment (danielsgriffin) [inline HW04_ch07_ex02.py]: `inputexpression = ''`

File: HW04_ch07_ex02.py [hunk 3]
```diff
@@ -21,15 +21,21 @@
 
 ###############################################################################
 # Imports
-
+import math
 
 # Body
-
+def eval_loop():
```
Comment (danielsgriffin) [inline HW04_ch07_ex02.py]: "return the value of the last expression it evaluated"

File: HW04_ch08_ex05.py [hunk 4]
```diff
@@ -2,3 +2,19 @@
 # See Chapter 8: Strings Exercise 5 for the task.
 # Please do provide function calls that test/demonstrate your
 # function.
+def rotate_word(word,shiftint):
```
Comment (danielsgriffin) [inline HW04_ch08_ex05.py]: Try: print(rotate_word('Z',1))

File: HW04_string2.py [hunk 5]
```diff
@@ -17,8 +17,13 @@
 # If the string length is less than 3, leave it unchanged.
 # Return the resulting string.
 def verbing(s):
-    # +++your code here+++
-    return
+    if len(s)>=3:
```
Comment (danielsgriffin) [inline HW04_string2.py]: add spaces: `if len(s) >= 3:`

Comment (danielsgriffin) [inline HW04_string2.py]: Instructions: "If the string length is less than 3"

File: HW04_string2.py [hunk 6]
```diff
@@ -17,8 +17,13 @@
...
+        if s[-3:]=='ing':
```
Comment (danielsgriffin) [inline HW04_string2.py]: `if s[-3:] == 'ing':`

**Comment by danielsgriffin** [comment]
Good job. (1) Watch the instructions. (2) Ensure you test edge cases (see comment for rotate_word)

File: HW04_ch08_ex05.py [hunk 7]
```diff
@@ -2,3 +2,19 @@
 # See Chapter 8: Strings Exercise 5 for the task.
 # Please do provide function calls that test/demonstrate your
```
Comment (danielsgriffin) [inline HW04_ch08_ex05.py]:
```
Book Examples:
 OK  'A'     by 3   => got: 'D'     expected: 'D'
  X  'Z'     by 1   => got: '['     expected: 'A'
 OK  'cheer' by 7   => got: 'jolly' expected: 'jolly'
 OK  'melon' by -10 => got: 'cubed' expected: 'cubed'
  X  'IBM'   by -1  => got: 'b[f'   expected: 'HAL'

Daniel's Tests:
 OK  'abc'   by 1   => got: 'bcd'   expected: 'bcd'
 OK  'xyz'   by 1   => got: 'yza'   expected: 'yza'
 OK  'ABC'   by 1   => got: 'BCD'   expected: 'BCD'
  X  'XYZ'   by 1   => got: 'YZ['   expected: 'YZA'
 OK  'Az'    by 1   => got: 'Ba'    expected: 'Ba'
  X  'Aa'    by 26  => got: '[a'    expected: 'Aa'
 OK  'abc'   by -1  => got: 'zab'   expected: 'zab'
 OK  'xyz'   by -1  => got: 'wxy'   expected: 'wxy'
  X  'ABC'   by -1  => got: 'Z[\\'  expected: 'ZAB'
  X  'XYZ'   by -1  => got: 'qrs'   expected: 'WXY'
 OK  'Az'    by -1  => got: 'Zy'    expected: 'Zy'
 OK  'Aa'    by -26 => got: 'Aa'    expected: 'Aa'
```

File: D04_ex00.py [hunk 8]
```diff
@@ -11,19 +11,40 @@
...
+        elif user_inp < rand_num and user_inp > 0:
+            print("Too low")
+            continue
+        else:
+            print("Number not in the range of 1 to 25")
+            continue
```
Comment (peterrowland) [inline D04_ex00.py]: I think you lose a guess if you guess outside of the range - but that could be treated as an error instead of a wrong guess.

File: HW04_ch08_ex05.py [hunk 9]
```diff
@@ -2,3 +2,19 @@
...
+    for s in word:
+        if shiftint>0 and ord(s)+shiftint>numz:
+            word1=word1+chr(ord(s)+shiftint-26)
+        elif shiftint<0 and ord(s)+shiftint<numa:
+            word1=word1+chr(ord(s)+shiftint+26)
+        else:
+            word1=word1+chr(ord(s)+shiftint)
+    return word1
+
+
```
Comment (peterrowland) [inline HW04_ch08_ex05.py]: You might be able handle both uppercase and lowercase wrapping using the same code you have, but adding a test at the beginning using `word.isupper()` and modifying the loop.

File: HW04_string2.py [hunk 10]
```diff
@@ -30,8 +35,12 @@ def verbing(s):
 # So 'This dinner is not that bad!' yields:
 # This dinner is good!
 def not_bad(s):
-    # +++your code here+++
-    return
+    notstr=s.find('not')
+    badstr=s.find('bad')
+    if notstr>0 and badstr>0 and notstr<badstr:
```
Comment (peterrowland) [inline HW04_string2.py]: I like this - all test conditions in one statement.

File: HW04_string2.py [hunk 11]
```diff
@@ -41,9 +50,10 @@ def not_bad(s):
 # e.g. 'abcde', the front half is 'abc', the back half 'de'.
 # Given 2 strings, a and b, return a string of the form
 #  a-front + b-front + a-back + b-back
+import math
+
 def front_back(a, b):
-    # +++your code here+++
-    return
+    return a[:math.ceil(len(a)/2)]+b[:math.ceil(len(b)/2)]+a[math.ceil(len(a)/2):]+b[math.ceil(len(b)/2):]
```
Comment (peterrowland) [inline HW04_string2.py]: .ceil was the way to go.
